### PR TITLE
tfruns integration, train_and_evaluate()

### PR DESCRIPTION
**Pull request by kevinykuo**
- `tfruns` integration
- Implement `train_and_evaluate()`
- Implement `plot()` and `print()` methods for training history objects

### CODE REVIEW

File: R/train_and_evaluate.R [hunk 1]
```diff
@@ -38,19 +38,46 @@
...
+      train_spec = do.call(tf$estimator$TrainSpec, train_spec$args),
+      eval_spec = do.call(tf$estimator$EvalSpec, eval_spec$args)
+    )
+  })
+  
+  training_history <- as.data.frame(.globals$history[[mode_keys()$TRAIN]])
+  steps_per_axis_unit <- training_history$steps[2] - training_history$steps[1]
+  
+  training_history <- if (nrow(training_history) > 100) {
+    # cap number of points plotted
```
Comment (kevinykuo) [inline R/train_and_evaluate.R]: @jjallaire FYI here's what motivated my question regarding time axis labels. The current solution is to cap the number of rows being written to the training history (and hence the number of points plotted in `tfruns`). I'm also leaving out the `steps` column so it doesn't get plotted, and writing a `steps_per_axis_unit` so users can interpret the plot (may need a new section in the `view_run()` dashboard to show this). Ideally we'd have actual step units on the time axis without blowing up the browser.

**Comment by jjallaire** [comment]
Yes I’m sure I can get this to work exactly how you want it to.

I think we should also have a plot method for the history object (perhaps
based on the code for the Keras history plot method)

File: R/train_and_evaluate.R (hunk 1 above)
Comment (jjallaire) [reply R/train_and_evaluate.R]: Should be addressed by https://github.com/rstudio/tfruns/commit/1d25ab63e8cfb196b30b3e5af0656e21f961d145

**Comment by kevinykuo** [comment]
@jjallaire The tfruns changes work perfectly, thanks!

Now, regarding the plot method -- the `train()` method in Python returns the trained estimator where we currently return a data frame (with the history), are we OK with deviating from Python behavior here? I think the one use case for returning `self` would be something like
```
model %>%
  train() %>%
  evaluate()
```
but it'd be nice if `plot(model)` works, too. Let me know if you have a preference. We could try hacking it with attributes or something, too...

**Comment by jjallaire** [comment]
In this case I think it's better to be consistent with Keras (as many Keras users will end using train also once we get all the TF datasets stuff working). I think you are right that the return is there for method chaining, however:

1) Scenarios for this seem limited (the `evaluate()` one you illustrated is the only one I can think of)

2) With objects that are modified in place we are trying hard in the R interface to **NOT** return them since that implicitly signals by-value semantics when there are none (obviously we do return them when method chaining is natural/obvious).

**Comment by kevinykuo** [comment]
@jjallaire could you take a look?

**Comment by jjallaire** [comment]
LGTM! I am going to merge and then play with the new history plotting and tfruns output

**Comment by jjallaire** [comment]
I get this when attempting to plot the history returned from train in the wide_and_deep.R example:

screen shot 2017-12-12 at 7 49 38 am

**Comment by jjallaire** [comment]
Here's the stack trace:

screen shot 2017-12-12 at 7 50 37 am

**Comment by jjallaire** [comment]
I get the same error when attempting to execute under `tfruns::training_run()` (the error in this case comes before `train()` does it's work)

**Pull request merged**